test/onnx/onnx_test_common.py

```diff
@@ -436,16 +436,18 @@ def _compare_pytorch_onnx_with_ort(
         ref_input_args = input_args
         ref_input_kwargs = input_kwargs
 
-    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(
-        ref_model(*ref_input_args, **ref_input_kwargs)
-    )
-
+    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
+    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
+    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
     ort_outputs = onnx_program(*input_args, **input_kwargs)
+    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
+    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```

Comment:
Why wasn't this a requirement? Could you elaborate more?

Reply:
there 2 changes here.

one is just aesthetic, not thing really changed. 
Before we had `    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_model(*ref_input_args,**ref_input_kwargs))` and now we have 
```python
    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```

The second change is swapping the order between `ref_outputs` and `ort_outputs` which is explained by the new comment:
```python
    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
```

Reply:
My question was why we didn't need the second change? Because having the need of the second change doesn't feel very intuitive.

Reply:
Offline discussion. When the input model is a `torch.nn.Module`, the FX tracer cannot handle buffer mutation, so the order in which ORT model and Torch model were never an issue. When the input model is `torch.export.ExportedProgram`, we internally use AOTAutograd which lifts mutated buffers as model inputs/outputs, so we need to execute ORT (with the pytorch model state) before torch to make sure both models are on the same iteration (forward step).